### ISSUE TITLE
Rotating server keypair

### ISSUE THREAD

**Issue by mitar** [cfrg/draft-irtf-cfrg-opaque#392]
I am missing mention/protocol/description of how can the server keypair be rotated. To me this is a bit surprising because I am used for keys to have expiration and are rotated before that expiration. OPAQUE seems to expect that server keys will be in use forever or at least be changed only after a compromise?

Currently, the only way I can see that rotation is done is by having support for that in clients so that they re-register after successful login. But clients might misbehave or simply skip that step. It looks to me like a potential deployment nightmare where some clients keep using old keypair.

**Comment by bytemare** [comment]
A client is not supposed to "keep using old keypairs": it fetches the server's public key on each run.

But indeed, when the server updates its AKE key pair, the client must re-register a new envelope.

**Comment by mitar** [comment]
> A client is not supposed to "keep using old keypairs": it fetches the server's public key on each run.

Hmmm. I must say this sounds a bit strange here. So public keypair is stored in the envelope client obtains from the server. Are you saying that it also additional fetches server's current public key? So what if they differ? How does it know if this means that the server rotated its key pair and the client should re-register a new envelope or that there is an attack and somebody is doing MITM? Or it does not matter?

> But indeed, when the server updates its AKE key pair, the client must re-register a new envelope.

So I think the question is how does the client know when to do so?

**Comment by bytemare** [comment]
> So public keypair is stored in the envelope client obtains from the server. Are you saying that it also additional fetches the server's current public key?

No, no key is stored in the envelope. The envelope holds a nonce and authentication tag. The authentication tag is computed, among other values, over the server's public key. The server sends its public key to the client alongside the envelope. It's almost the same. But yes, through this the server commits to using the same key.

> So I think the question is how does the client know when to do so?

Since that is an application-specific problem, it would be up to the application to inform the client.
